lib/pf/mfa/Akamai.pm

```diff
@@ -0,0 +1,507 @@
+            @device = grep { $_->{'default'} == 1} @{$devices};
+        }
+        # Return the n-1 phone
+        @device = @{$devices}[$phone_id-1];
+    } else {
+        # Return the default phone
+        @device = grep { $_->{'default'} == 1 } @{$devices};
+    }
+    return @device;
+}
+
+=head2 totp
+
+totp method
+
+=cut
+
+sub totp {
+    my ($self, $device, $username, $otp) = @_;
+    my $logger = get_logger();
+    my $method = "offline_otp";
+    if (length($otp) == 16) {
+        $method = "bypass_code";
+    }
+    $logger->info("Trigger $method for user $username");
+    my $post_fields = encode_json({device => $device, method => { $method => {"code" => $otp} } , username => $username});
+    my ($auth, $error) = $self->_post_curl("/api/v1/verify/start_auth", $post_fields);
+    if ($error) {
+        return $FALSE;
+    }
+    if ($auth->{'result'}->{'status'} eq 'allow') {
+        $logger->warn("Authentication sucessfull on Akamai MFA");
+        return $TRUE;
+    }
+    $logger->warn("Authentication denied on Akamai MFA, reason: ". $auth->{'result'}->{'status'}->{'deny'}->{'reason'});
+    return $FALSE;
+}
+
+=head2 generic_method
+
+generic method
+
+=cut
+
+sub generic_method {
+    my ($self, $device, $username, $method) =@_;
+    my $logger = get_logger();
+    $logger->info("Trigger $method for user $username");
+    my $post_fields = encode_json({device => $device, method => $METHOD_LOOKUP{$method}, username => $username});
+    my $chi = pf::CHI->new(namespace => 'mfa');
+    my ($auth, $error)= $chi->compute($device.$METHOD_LOOKUP{$method}, {expires_in => normalize_time($self->cache_duration)}, sub {
+            return $self->_post_curl("/api/v1/verify/start_auth", $post_fields);
+        }
+    );
+    if ($error) {
+        return $FALSE;
+    }
+    # Cache the method to fetch it on the 2nd radius request (TODO: cache expiration should be in config).
+    if (!$chi->get($username)) {
+        my $infos = { username => $username,
+                      device   => $device,
+                      tx       => $auth->{'result'}->{'tx'},
+                    };
+        $chi->set($username, $infos, normalize_time($self->cache_duration));
+    }
+    # Remove the authenticated status of the user since the next radius requests will use OTP
+    $chi->remove($username." authenticated");
+    return $FALSE;
+}
+
+=head2 push
+
+push method
+
+=cut
+
+sub push {
+    my ($self, $device, $username) =@_;
+    my $logger = get_logger();
+    $logger->info("Trigger push for user $username");
+    my $post_fields = encode_json({device => $device, method => "push", username => $username});
+    my $chi = pf::CHI->new(namespace => 'mfa');
+    my ($auth, $error)= $chi->compute($device."push", {expires_in => normalize_time($self->cache_duration)}, sub {
+            return $self->_post_curl("/api/v1/verify/start_auth", $post_fields);
+        }
+    );
+    if ($error) {
+        return
+    }
+    my $i = 0;
+    while(1) {
```

Comment:
This could technically be stuck for a very long time if the check_auth call experiences a delay

We should instead use an alarm instead to limit the amount of time this can take